## 1.1 Instalación del software necesario, entorno de desarrollo y primeros pasos ##

#### MATLAB y Octave ####

MATLAB (abreviatura de MATrix LABoratory, 'laboratorio de matrices') es un software matemático de cálculo técnico y científico. Fue creado por Cleve Moler en 1984.

MATLAB es un programa especialmente diseñado para efectuar cálculos numéricos con vectores y matrices, que resultan fundamentales a fin de resolver problemas de ingeniería y ciencia. Además, una de sus características más atractivas es la de facilitar la realización de una amplia variedad de gráficos en dos y tres dimensiones. MATLAB posee también un lenguaje de programación propio, de muy alto nivel (lenguaje M). En definitiva, es una magnífica herramienta para desarrollar aplicaciones técnicas, muy utilizada en el campo de la ingeniería
y fácil de manejar.

Pueden extender las capacidades del módulo básico de MATLAB con las cajas de herramientas (toolboxes) que consisten en paquetes de ampliación del software básico y que son aplicables a determinados campos de la ciencia: matemática en general (optimización, bases de datos, estadística, ecuaciones diferenciales, redes neuronales), procesado de señales, procesado de imágenes, adquisición de datos, finanzas, sistemas de control, etc.

MATLAB es un software abierto que puede relacionarse con otras aplicaciones, como Excel, C,
Fortran, etc. A modo de ejemplo, los programas escritos en MATLAB pueden ser traducidos a
lenguaje C de forma inmediata. En este curso nos basaremos en el uso del software básico de
MATLAB.

Octave o GNU Octave es un software libre para realizar cálculos numéricos que también posee un lenguaje de programación propio. Su versión 1.0 apareció en 1994. Es considerado el equivalente libre de MATLAB, ya que aunque existen ciertas diferencias sintácticas, la mayoría son debidas a que Octave es más amplio y admite alguna sintaxis que MATLAB no permite. Luego si se utiliza sintaxis MATLAB los programas van a funcionar en ambos software.

Para seguir este curso, se puede utilizar indistintamente el software MATLAB u Octave. Hablaremos en este curso de lenguaje M para referirnos, indistintamente, al lenguaje que se utiliza en ambos software.

#### Instalación del software necesario para el seguimiento del curso ####

MATLAB es un software comercial, por tanto, con coste. El alumno que lo tenga instalado en su computador puede seguir el curso con este programa.

Sin embargo, el curso también puede seguirse usando el software libre, gratuito, Octave. Este software se puede descargar desde (https://www.gnu.org/software/octave/download.html ). En el vídeo explicativo correspondiente a este tema se indica la forma en la que se debe realizar la instalación de este.

Para continuar con el curso el alumno debe tener instalado en su computador, bien el software Octave, o bien el software MATLAB (en cualquiera de las versiones).

#### El entorno de desarrollo integrado en MATLAB y Octave ####

Para poder escribir y ejecutar un programa en un lenguaje de programación se necesita tener instalado en el computador un compilador o intérprete para tal lenguaje. MATLAB y Octave utilizan un intérprete.

El intérprete no se usa de manera aislada. Suele incluirse en Entornos de Desarrollo Integrados (IDE, por sus siglas en inglés). El entorno que vamos a usar cuenta con:

- Editor de código fuente (Editor)
- Explorador de variables (Workspace).
- Línea de comandos del intérprete (Command Window).
- Historial de comandos (Command History)
- Explorador de ficheros (Current Folder)
- Depurador (Debug)
- Otras herramientas

Sea cual sea la versión de MATLAB que utilicemos o si estamos usando Octave lo verdaderamente importante es saber que existen estas herramientas y utilizarlas para potenciar y acelerar nuestro trabajo.

A continuación indicamos la función principal de las ventanas y/o herramientas principales del IDE:

- Editor, es el editor de texto en el que se escribirá el código fuente de los programas, además de poder visionar cualquier archivo de texto plano (texto sin formato).

- Workspace, guarda la información de las variables utilizadas en la sesión de trabajo actual.
- Command Window, es la ventana más importante, ya que en ella aparece la línea de comandos del intérprete, en la que se teclean las instrucciones a ejecutar para obtener un resultado de inmediato.

- Command History, almacena todas las sentencias que se han ejecutado en la ventana de comandos en las últimas sesiones de trabajo. Se puede navegar a través de ellas mediante la flecha de desplazamiento vertical pulsada desde Command Window.

- Current Folder indica cuál es el directorio de trabajo y los archivos y/o directorios incluidos en él.

- Debug, herramienta que permite actuar en la ejecución del programa mediante múltiples acciones: insertar puntos de parada en el código, ejecutar el código paso a paso, etc.

## 1.2 Uso del software como una calculadora. Iniciación a los operadores y almacenamiento de datos ##
#### Iniciación al manejo de datos ####
Empezaremos a usar el software en modo calculadora, utilizando la ventana de
comandos para escribir la instrucción y observando el resultado.

In [1]:
2+3

ans =  5


Todo resultado de una operación se debe almacenar en la memoria del computador. Si
no se indica explícitamente el nombre de la variable donde se guardará éste, el lenguaje M usa
la variable reservada *ans*.

Podemos por tanto interpretar que el resultado de la operación $2+3$ se ha almacenado
en la variable *ans*.

Cuando el programador use sus propias variables debe utilizar nombres válidos. Como
identificador (nombre) se puede utilizar cualquier palabra que empiece por una letra y tenga a
continuación cualquier combinación de letras, números o el carácter guión bajo. Es importante
conocer que se distinguen mayúsculas de minúsculas, luego, por ejemplo, las variables $A$ y $a$
son distintas.

Otra cosa importante es saber cómo introducir los valores numéricos en las sentencias de este
lenguaje. Los valores numéricos enteros se escriben literalmente, y los valores reales, usando
un punto para separar la parte entera de los decimales o también utilizando notación
científica. La notación científica es una forma de representar un número utilizando potencias
de 10. Por ejemplo, el valor 34.2865 se podría escribir en notación científica como
0.342865x10 2 (se escribirá como 0.342865e2). Veamos un ejemplo en el que asignamos un
valor entero a la variable $a$, mientras que a las variables $c$ y $d$ se les asignan valores reales

In [2]:
a=23, c=45.876, d=1.453e-2

a =  23
c =  45.876
d =  0.014530


`M` tiene variables predefinidas que contienen un valor por defecto. Por ejemplo, entre otras,
existe la variable *pi* con el valor correspondiente a la constante matemática $\pi$.

Existen algunas variables que representan cantidades que no son consideradas números como
tal. Estas cantidades son:
- Infinito, $\infty$, representado por la variable Inf (o inf)
- Cantidades indefinidas, representadas por la variable NaN (o nan)

In [3]:
Inf # Equivalentes
inf # Equivalentes
NaN # Not a number (0/0, \inf-\inf)
nan

ans = Inf
ans = Inf
ans = NaN
ans = NaN


#### Operador de asignación ####

El operador de asignación (=) se utiliza para dar un valor nuevo a una variable. Su utilización es la siguiente:

</p>
<p align="center">
  <b>*variable = expresión*</b>
  <br>
</p>

donde *expresión* es cualquier expresión válida en lenguaje M.

Si la variable no existe, ésta se creará con el valor y el tipo de dato de la expresión situada a la
derecha del operador de asignación.

Si la variable existe, cada vez que se le asigne un nuevo valor, ésta pierde su estado anterior
(tipo de dato y valor) y toma un nuevo estado correspondiente con el valor y el tipo de la
expresión de la derecha.

#### Operadores aritméticos ####

Los operadores aritméticos aplicables a escalares (ampliaremos éstos cuando se
estudien matrices) son los siguientes: Suma (+), resta (-), producto (*), división derecha (/), división izquierda (\), potenciación (^).

Resto de una división: *no existe un operador*. A tal efecto se debe utilizar la función
`rem(A,B)` que calcula el resto de dividir $A$ entre $B$.

Para desarrollar una sentencia en la que aparezcan varios operadores, las operaciones se
deben realizar siguiendo la prioridad de operadores. La siguiente tabla muestra la prioridad de
operadores aritméticos y asignación de mayor a menor prioridad.

|**Operador**|**Símbolos**|**A igual prioridad se sigue el orden de**|
|:-----:|:------:|:---------:|
|Paréntesis| ()||
|Potencia| ^| Izquierda a derecha|
|Multiplicación, división| *,/,\| Izquierda a derecha|
|Suma y resta, cambio de signo(-)|+,-|Izquierda a derecha|
|Asignación|=|Derecha a izquierda|

In [4]:
2.4*6
2^3
c=-1^4
3/4
3\4
2/3^2
x=2/3*2
b=rem(6.7,5)

ans =  14.400
ans =  8
c = -1
ans =  0.75000
ans =  1.3333
ans =  0.22222
x =  1.3333
b =  1.7000


#### Formatos para el muestreo de resultados en pantalla ####

Independientemente de la precisión con la que un dato se ha memorizado en la
memoria RAM, podemos elegir entre diferentes formatos de escritura de los datos en pantalla.
Existen varios modos de trabajo en MATLAB y Octave. Se indican a continuación, los más
significativos:
- `format short`: punto fijo con cinco cifras significativas 1 . Si el dato no puede mostrarse adecuadamente con este formato opta por un formato en coma flotante.
- `format long`: punto fijo con quince cifras significativas. Si el dato no puede mostrarse adecuadamente con este formato opta por un formato en coma flotante.
- `format short e`: formato en coma flotante. Para representar la mantisa se usan cinco cifras significativas.
- `format long e`: formato en coma flotante. Para representar la mantisa se usan quince cifras significativas.
- `format rat`: aproximación por formato racional (cociente de enteros)

El modo de trabajo por defecto en lenguaje M es `format short` luego los datos se mostrarán con cinco cifras significativas mientras que el usuario no quiera modificar el formato.

A continuación se utiliza MATLAB, como si de una calculadora se tratara, para calcular los valores $3^100$ y $5+\pi$ , cambiando el formato de muestreo de resultados. (Adelantamos aquí el uso del operador potenciación (^)).

In [5]:
# format short
3^100
5+pi
format short e
3^100
5+pi
format long
3^100
5+pi
format long e
3^100
5+pi
format rat
5+pi

ans =    5.1538e+47
ans =  8.1416
ans =    5.1538e+47
ans =    8.1416e+00
ans =    5.15377520732011e+47
ans =  8.14159265358979
ans =    5.15377520732011e+47
ans =    8.14159265358979e+00
ans = 920/113


Tecleando `format` se vuelve al formato por defecto, es decir, `format short`.

Usemos ahora la ventana de comandos para ver un ejemplo de los errores en el almacenamiento de datos. Escribamos en la ventana de comandos el número $2$, y a
continuación el resultado de la operación ${(\sqrt{2})}^2$ (Adelantamos aquí el uso de la función raíz
cuadrada (`sqrt()`). Mostremos 15 cifras significativas.

In [6]:
2
format long
(sqrt(2))^2

ans = 2
ans =  2.00000000000000


Matemáticamente ambos resultados son iguales y computacionalmente parecen iguales. Vamos a comprobarlo:

In [7]:
2-(sqrt(2))^2

ans =   -4.44089209850063e-16


El resultado debería ser cero, pero como el ordenador ha tenido que redondear $\sqrt{2}$ ya que
tiene infinitos decimales, arrastra este error y la diferencia resulta distinta de cero. Se observa
que hay diferencia a partir del decimal número 16, puede parecer un error pequeño, pero si no
se tiene en cuenta, puede ocurrir que:
- si estamos comparando dos cantidades, la diferencia nunca puede ser cero y el programa fallará.
- si estamos realizando un cálculo de ingeniería, se puede producir una propagación y acumulación de errores que dé lugar a un resultado final con un error inaceptable.

Podemos averiguar cuál es la precisión relativa en punto flotante usando el comando `eps`

In [8]:
eps

ans =    2.22044604925031e-16


In [18]:
format
pi
eps
realmax
realmin

ans =  3.1416
ans =    2.2204e-16
ans =   1.7977e+308
ans =   2.2251e-308


Esto significa que si tenemos el valor 1 guardado, el siguiente valor que podemos almacenar es
1+eps. Es debido a que en coma flotante y doble precisión se tienen 52 bits para la mantisa,
siendo `eps`=$2^{-52}$

## 1.3 Tipos de datos elementales, operadores y comandos utilitarios ##
#### Tipos de datos elementales ####

M es fundamentalmente un lenguaje para cálculo matricial. Todos los datos que maneja son matrices, pudiendo también trabajar con vectores y escalares, pero considerando a éstos casos particulares de matrices: un vector fila es una matriz de una fila; un vector columna, es una matriz de una columna; un escalar es una matriz de una fila y una columna.

El lenguaje M, en general, trabaja con datos numéricos, utilizando, por defecto, para su almacenamiento doble precisión (tipo de dato `double`), es decir, manejando 8 bytes de memoria para cada dato, con 15 cifras significativas en punto flotante. También puede trabajar con otros tipos de datos como enteros, reales de simple precisión, datos lógicos, cadenas de caracteres y con tipos de datos más avanzados: hipermatrices, estructuras, matrices de celdas y clases y objetos.

Combinando el tipo de dato por defecto (`double`) junto con datos lógicos y de carácter podría realizarse casi cualquier programa, por ello, son los tipos de datos que manejaremos en este curso.

#### Datos numéricos double ####

Las variables reales en doble precisión (tipo `double`), son las variables por defecto
que maneja el lenguaje M. Utilizan 8 bytes en memoria para almacenar el dato. No hay que
hacer ninguna declaración, simplemente se usan. Los valores máximos y mínimos que se
pueden almacenar son `1.7977e+308` y `2.2251e-308`. Sin embargo, la precisión de los
cálculos no es superior a 15 cifras significativas.

#### Datos de tipo lógico ####

Es habitual al trabajar con cualquier lenguaje de programación que aparezcan datos de contenido lógico (true o false). Estos datos se manejan en lenguaje M con los valores 1 o true (cierto) y 0 o false (falso).

Lo normal es que estos datos se generen automáticamente como resultado de ciertas operaciones.

Si un valor no lógico se tiene que evaluar como lógico se hará la transformación de la siguiente manera:

- Cualquier dato distinto de 0 equivale a cierto (1).
- Cualquier dato igual a cero equivale a falso.

#### Cadenas de caracteres ####

Además de datos numéricos es necesario manejar texto. M puede definir variables que contengan cadenas de caracteres, para ello las cadenas de texto se deben delimitar entre apóstrofos o comillas simples. En el siguiente ejemplo se define la variable s que contiene un texto.

In [9]:
a='Esto es una cadena'

a = Esto es una cadena


M maneja la cadena como una matriz de una fila (vector fila) distinguiendo cada carácter como si fueran distintas componentes del vector. Así, en este caso tenemos un vector de 18
componentes.

Veamos un fragmento de programa en el que se puede observar la necesidad de entrecomillar un valor de carácter para diferenciarlo del nombre de una variable.

In [10]:
a=7;
v=a;
w='a'

w = a


Fíjense que la variable v contiene el valor 7 y la variable w el carácter 'a'.

#### Otros Operadores ####

###### Operadores relacionales y de igualdad ######

Relacionan dos valores numéricos y dan como resultado un valor lógico (cierto (1) o falso (0)). Se muestran en la siguiente tabla:

|     |    |
|:---:|:---:|
|Menor que|`<`|
|Mayor que|`>`|
|Menor o igual que|`<=`|
|Mayor o igual que|`>=`|
|Equivale a |`==`|
|Distinto de|`~=`|

Ejemplos de utilización de estos operadores:

In [11]:
u=3.5;
v=7.3;
w=7.3;
u<v
v>w
v>=w
u~=v
v==w

ans = 1
ans = 0
ans = 1
ans = 1
ans = 1


#### Operadores lógicos ####

Se aplican a valores lógicos resultando otro valor lógico. Son los siguientes:

##### Operador lógico OR:#####

Se representa con el carácter `|`. Se utiliza entre dos valores lógicos
dando como resultado cierto, si ambos o uno de ellos son ciertos. Sólo si los dos son
falsos resulta falso. La función equivalente es `or(A,B)`.

Ejemplo: Si la calificación de un examen se guarda en la variable `nota`, la expresión que resulta cierta si la nota es errónea es: `nota<0|nota>10`.

##### Operador lógico AND: ######

Se representa con el carácter &. Se utiliza entre dos valores
lógicos dando como resultado cierto sólo si ambos son ciertos. Si uno o los dos son
falsos el resultado es falso. La función equivalente es `and(A,B)`.

Ejemplo: Si la calificación de un examen se guarda en la variable `nota`, la expresión
que resulta cierta si la nota es igual a notable es: `nota<9&nota>=7`.

##### Operador lógico OR EXCLUSIVO: #####

Se utiliza con la sintaxis `xor(A,B)`, resultando cierto cuando A o B son ciertos, pero no ambos.

Ejemplo: Supongamos que un alumno realiza dos pruebas cuyas calificaciones se guardan en las variables `nota1` y `nota2`, y se ofrece una recuperación a los alumnos que hayan suspendido sólo una de las dos, la expresión que resulta cierta en ese caso sería: `xor(nota1<5,nota2<5)`.

##### Operador lógico NOT: #####

Se representa con el carácter `~`. Actúa sobre el valor lógico situado a su derecha, resultando el valor lógico contrario a éste. La función equivalente es `not(A)`. Suele utilizarse para hacer referencia a una condición contraria a la que se esté estudiando.

Ejemplo: Si la calificación de un examen se guarda en la variable `nota`, la expresión que resulta cierta si la nota es suspensa es: `~(nota>=5)`.

##### Operadores lógicos BREVES (&&) y (||) : #####

Son realmente el `AND` y el `OR` pero permiten simplificar la operación. Veamos las diferencias:

`A&B`: estudia siempre las condiciones A y B.
`A&&B`: estudia la condición A, si es cierta estudia la B, pero si es falsa ya no estudia la B, porque el resultado final se conoce que es falso.
`A|B`: estudia siempre las condiciones A y B.
`A||B`: estudia la condición A, si es falsa estudia la B, pero si es cierta ya no estudia la B, porque el resultado final se conoce que es cierto.

Esto además de simplificar los cálculos se puede utilizar para evitar posibles errores. Por ejemplo, en la expresión:

In [12]:
c= a~=0&&b/a>5,

c = 0


si el valor de la variable a es cero, ya no se desarrolla la segunda expresión y así se evita el error que supondría ésta (división entre cero).

#### Prioridad de operadores ####
En la tabla siguiente se muestra el orden de prioridad todos los operadores conocidos hasta el momento. Se han ordenado de mayor a menor prioridad (de arriba hacia abajo).

|**Operador**|**Símbolos**|**A igualdad prioridad se sigue el orden de**|
|:----------:|:----------:|:-------------------------------------------:|
|Paréntesis, or exclusivo| | |
|No lógico| |Derecha a izquierda|
|Potencia||Izquierda a derecha|
|Multiplicación, división||Izquierda a derecha|
|Suma y resta, y cambio de signo (-)||Izquierda a derecha|
|Operador :||Izquierda a derecha|
|Relacionales||Izquierda a derecha|
|Igualdad||Izquierda a derecha|
|Y lógico||Izquierda a derecha|
|O lógico||Izquierda a derecha|
|Asignación||Derecha a izquierda|

#### Comandos utilitarios en el manejo del lenguaje M ####

##### Almacenamiento del texto de la ventana workspace #####

Si se desea almacenar en un fichero de texto las entradas que se van a teclear en la ventana de comandos y las salidas ofrecidas por el programa, lo más sencillo es utilizar el comando `diary`.

Tecleando

</p>
<p align="center">
  <b>`diary nombrefichero`</b>
  <br>
</p>

se crea el fichero con el nombre indicado en la ubicación actual almacenándose en él el
contenido que se inserte en la ventana *workspace* a partir de ese momento. Si no se indicara
nombre de fichero, se crearía el archivo de nombre diary para tal fin. Si en un momento de la
sesión se quiere desactivar la grabación en el fichero, se escribirá

</p>
<p align="center">
  <b>`diary off`</b>
  <br>
</p>

cuando quiera reanudarse la grabación, se tecleará

</p>
<p align="center">
  <b>`diary on`</b>
  <br>
</p>

También se puede pasar del estado *on* al *off*, o viceversa, escribiendo simplemente `diary`.

#### Almacenamiento de variables en ficheros binarios ####

Cuando se finalice una sesión de trabajo, para evitar perder los datos obtenidos (variables) se pueden guardar éstos (con sus valores) dentro de un fichero binario, que puede ser cargado en otro momento para continuar con el trabajo anterior. Esta operación se puede realizar utilizando el comando `save`. A continuación se ejemplifican diferentes opciones de utilización.

Si se quieren guardar todas las variables del espacio de trabajo en el fichero de nombre *guardavariables*, se utilizaría:

</p>
<p align="center">
  <b>`save guardavariables`</b>
  <br>
</p>

Si sólo interesa guardar algunas variables, por ejemplo, las variables de identificadores $w$ y $z$, se procede así:

</p>
<p align="center">
  <b>`save guardavariables w z`</b>
  <br>
</p>

Para cargar de nuevo las variables guardadas en un fichero binario se utiliza el comando `load` de la forma siguiente:

</p>
<p align="center">
  <b>`load guardavariables`</b>
  <br>
</p>

#### Otros comandos útiles ####

##### Comentarios en el código #####

Los comentarios en el código fuente de un programa se añaden para hacer el código más entendible al programador de cara a futuras utilizaciones o a compartir el código con terceros. Se trata de texto que se incrusta en el código pero que es ignorado por el compilador o intérprete.

En lenguaje M, para insertar un comentario debemos preceder éste por el símbolo %. En el siguiente ejemplo, la primera línea es un comentario, también aparece un comentario después de la instrucción de la tercera línea.

In [13]:
% radio y altura del cilindro
r=7.3;h=25.4;
a=pi*r^2*h; %volumen del cilindro

##### Continuación de una sentencia en otra línea #####

En ocasiones la escritura de una sentencia queda demasiado larga y conviene continuarla en la línea posterior. Para ello se debe terminar la línea que continuará en la siguiente con tres puntos. Veamos un ejemplo:

In [14]:
b=7;c=6;
x=b+c^3-c*b...
+4/b

x =  181.571428571429


Sin embargo no se permite realizar este proceso en una cadena de caracteres. El siguiente ejemplo produciría un error:

In [15]:
%x='No es posible continuar un texto...
%en la siguiente línea'

##### Eliminación de variables del espacio de trabajo #####

Si se desea eliminar todas las variables del espacio de trabajo, se utiliza el comando `clear`. Para eliminar únicamente algunas variable se emplea el mismo comando seguido de los nombres de las variables a eliminar. Por ejemplo la siguiente instrucción borra las variables $a$, $b$ y $c$.

</p>
<p align="center">
  <b>`clear a b c`</b>
  <br>
</p>

#### Limpieza de texto de la ventana de comandos ####

Para eliminar todo el texto escrito en la ventana de comandos se utiliza el comando `clc`. La ejecución de esta orden no afecta a las variables de la sesión de trabajo (la ventana workspace sigue manteniendo las variables).

#### Directorio de trabajo ####

Para conocer la ruta al directorio de trabajo actual se utiliza el comando `pwd`.

Para obtener un listado de los ficheros que existen en la carpeta actual, se utiliza el comando `dir`.

Para cambiar de carpeta de trabajo se utiliza el comando:

</p>
<p align="center">
  <b>`cd nombrenuevacarpeta.`</b>
  <br>
</p>

#### Utilización de la ayuda ####

Para obtener ayuda sobre la utilización de un comando o función determinada se utiliza la sentencia:

</p>
<p align="center">
  <b>`help nombredelcomando`</b>
  <br>
</p>

Para buscar un texto que esté contenido en la primera línea de comentario de los programas se utiliza el comando `lookfor` de la forma:

</p>
<p align="center">
  <b>`lookfor 'texto a buscar'`</b>
  <br>
</p>

#### Tiempo que tarda un cálculo en ser efectuado #####

Cuando se implementa un algoritmo en un lenguaje de programación es habitual que se pueda obtener la solución esperada mediante diferentes códigos fuente. Una forma de comparar los códigos y elegir cuál resulta más conveniente es calcular el tiempo de ejecución. El lenguaje M cuenta con las funciones tic y toc que trabajan juntas y calculan el tiempo transcurrido entre la ejecución de la primera y la segunda.

`tic`: activa un contador temporal en segundos que finaliza al utilizar el comando `toc`.
`toc`: devuelve el tiempo transcurrido en segundos desde que se activó el contador con `tic`.

Veamos unos ejemplos que muestran dos formas de utilización. En ambas se almacena el tiempo de ejecución en la variable `tiempo`:

In [16]:
tic
suma=0;
for i=1:1000
suma=suma+i;
end
tiempo=toc

tiempo =  0.0150940418243408


In [17]:
clc, clear all;
inicio=tic
suma=0;
for i=1:1000
suma=suma+i;
end
tiempo=toc(inicio)

inicio = 1516300920186517
tiempo =  0.0171461105346680


## 1.4 Ficheros m. Entrada y salida de datos por consola. ##

#### Ficheros m ####

Hasta el momento todas las órdenes del lenguaje M las hemos ejecutado desde la ventana de comandos. De esta manera obtenemos una respuesta inmediata a la instrucción. Sin embargo si queremos realizar un proceso formado por una secuencia de instrucciones M que vamos tecleando una a una en la ventana de comandos:

- ¿qué ocurre si cometemos una imprecisión en una instrucción a lo largo del proceso? Habría que volver a realizarlo de nuevo, orden a orden, escribiendo esta vez sin cometer errores.

- ¿qué ocurre si se quiere volver a repetir este proceso en el futuro? Habría que volver a realizarlo de nuevo, ya que las instrucciones de la ventana de comandos no se pueden guardar como tal.

Estamos de acuerdo que esta forma de actuar no es eficiente. Para resolver estos problemas existen los ficheros M. Estos ficheros no son más que archivos de texto en los que se escriben las instrucciones, en lugar de en la ventana de comandos, que quedan guardados de forma permanente, y después pueden ser ejecutados por MATLAB u Octave. Los programas en M se escriben en ficheros de extensión .m (ficheros M) con un editor especial situado dentro del entorno de MATLAB u Octave, aunque es válido cualquier editor de texto no formateado, almacenándose en disco con un nombre cualquiera, por ejemplo, programa.m.

Para ejecutar el programa desde la ventana de comandos simplemente se debe teclear el nombre del fichero M que contiene el programa (sin la extensión). En el caso anterior sería:

>> programa

Entonces, se van ejecutando de manera secuencial todas las órdenes escritas en el fichero M, obteniendo los resultados en la ventana de comandos.

A modo de ejemplo, el alumno puede probar a teclear una a una las siguientes instrucciones en la ventana de comandos:

In [1]:
# Guardar script como ejemplo1.m
A=3
c=A^3
B=A+c

A =  3
c =  27
B =  30


y guardémoslo como ejemplo1.m

Se ejecuta desde la ventana de comandos:

>>A=3
>>c=A^3
>>B=A+c

Imaginemos ahora que la variable A se ha tecleado incorrectamente, ya que queríamos escribir

>> A=13

¿Cómo se resolvería esto desde la ventana de comandos? Evidentemente habría que volver a teclear las tres órdenes.

Probemos ahora a generar un archivo .m. secuencia:

O desde MATLAB 2016:

Escribamos las tres instrucciones en él:

y guardémoslo como ejemplo1.m
Se ejecuta desde la ventana de comandos:

>> ejemplo1

obteniéndose el resultado.

En este caso la corrección del error en la variable A se resolvería corrigiendo el valor correspondiente en el fichero y volviendo a ejecutar éste. Además queda grabado en disco, luego puede ejecutarse en cualquier otro momento.

Podemos decir que hemos realizado nuestro primer programa M.

A partir de este momento se trabajará con programas, es decir escribiendo las instrucciones en ficheros M. Poco a poco se irá enriqueciendo el contenido de estos programas.

El objetivo del resto de este tema es conseguir que nuestros programas puedan tomar, en tiempo de ejecución, valores que introduce el usuario por teclado (entrada de datos) y ofrecer respuestas a través de la ventana de comandos.

#### Entrada de datos por teclado ####

La función `input` permite introducir datos en un programa cuando está en modo de ejecución. La utilización es como sigue:

</p>
<p align="center">
  <b>`v=input('Texto de petición del dato')`</b>
  <br>
</p>

`input` realiza dos tareas:

    1) hace que aparezca en pantalla la cadena de caracteres que lleva como argumento.

    2) espera a que se tecleen los datos como respuesta al texto y los memoriza en la variable v.
    
`input` no permite disponer varias variables a la izquierda de la asignación, luego todos los datos que se introduzcan deben constituir una única variable con sintaxis matricial. Ejemplo:

In [2]:
P=input('Introduce el radio del círculo: ')

Introduce el radio del círculo:1.47
P =  1.4700


Si como respuesta a `input` se quiere introducir una cadena de caracteres, el usuario del
programa debe introducirla entre apóstrofos. El olvido de éstos provoca el fallo del programa.

In [4]:
Nombre=input('Introduce tu nombre y apellidos: ')

Introduce tu nombre y apellidos:Pedro Perez
error: 'Pedro' undefined near line 1 column 1
error: called from
    input at line 43 column 9


Para prevenirlo, se usa la función `input` con un segundo argumento '`s`' que hace que el dato introducido se tome como una cadena de caracteres sin necesidad de delimitarlo por los apóstrofos.

In [7]:
Nombre=input('Introduce tu nombre y apellidos: ','s')
%La variable Nombre contiene una cadena de caracteres

Introduce tu nombre y apellidos:Juan Pérez González
Nombre = Juan Pérez González


A veces un programa puede fallar porque el usuario no responda a la petición del dato y pulse simplemente la tecla *enter*. Para detectarlo puede ser de utilidad la función isempty que
devuelve cierto si la variable que tiene como argumento está vacía. Véase el ejemplo en el módulo 3 -comandos repetitivos-.

## 1.5 Practicando con programas(I). Ejercicios de autoevaluación ##

En esta lección se pretende que el alumno simplemente practique realizando los ejercicios que se proponen a continuación. Después de los enunciados de los ejercicios se encuentran las soluciones para que en caso de dificultad, en estos primeros acercamientos a la programación, se puedan analizar los programas bien planteados.

- Ejercicio 1.5.1 Escribir un programa M que pida al usuario el radio de un círculo y escriba en pantalla el área del mismo, que se mostrará con tres decimales, de la forma:

``
El área es xx.xxxx
``

In [5]:
clc, clear all;
radio = input('Ingrese el radio del círculo en metros: ');
fprintf('El área del círculo es %.3f', pi * radio * radio)

Ingrese el radio del círculo en metros:10
El área del círculo es 314.159
El área del círculo es 314.159
